replace mkYarnModules with yarnSetupHook

yu-re-ka opened the pull request:
###### Motivation for this change

The initial aim was to get rid of the big `yarn.lock` files that we still have to vendor because upstream has inconsistent lockfiles checked in, for example in hedgedoc. I would much rather have only a patch than the whole lockfile.

While looking into how we could inherit `src` and `patches` to `mkYarnModules`, I noticed that `mkYarnModules` doesn't have a `src` parameter at all, so using patches with it would be difficult.

I also noticed that that a dozend packages don't use mkYarnPackage, but instead run `yarn install` in the main derivation. 
All of this inspired me to create a setup hook that uses an offlineCache produced by either yarn2nix or fetchYarnDeps to run `yarn install` and create a node_modules folder in the main derivation. This has many advantages:
- eliminates one intermediate derivation, saving store space
- allows for much better support for patching `yarn.lock`
- better support for `path:` dependencies when they are included in src
- makes the node_modules folder writable, removing the need to create a writable copy in the working directory
- Can remove duplicated code in GitLab, schildichat-*, etc. by using the setup hook
- This gets rid of a lot of complicated logic like `pkgConfig`, which we only use twice in nixpkgs and which can easily be replaced

Potential disadvantages:
- `yarn install` is run each time the package changes, even if the lock file stays the same => I don't think this is a big disadvantage, because yarn install is quite fast
- Some of the advanced features of yarn2nix might not work properly without the complicated logic => I don't think that many people used it, or that it would be hard to replace with custom postConfigure hooks

All packages currently using mkYarnModules (not mkYarnPackage, I count 4 packages) would have to switch to the new yarnSetupHook.

Please tell me what you think about this direction in general.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

yuyuyureka committed:
fetchYarnDeps: support more url types

yuyuyureka committed:
yarnSetupHook: init

yuyuyureka committed:
hedgedoc: use new possibility to have patches for yarn.lock

yuyuyureka committed:
element-desktop: fix build after mkYarnPackage changes

yuyuyureka committed:
gitlab: use yarnSetupHook

happysalada commented on pkgs/servers/web-apps/plausible/default.nix:
```diff
@@ -36,26 +37,11 @@ let
     ];
   };
 
-  yarnDeps = mkYarnModules {
```
Just to let you know. I have a pending PR on plausible to update it to 1.4.0, it's ready but needs approval. If you don't mind, you'll need to wait a bit until it's merged then you'll be able to rebase.

yu-re-ka replied:
I am aware of that, thanks :)

happysalada commented on pkgs/servers/web-apps/plausible/package.json:
```diff
@@ -1,57 +0,0 @@
-{
-  "repository": {},
-  "license": "MIT",
```
In the last version of plausible, the package.json is missing version number, it was causing problems with yarn.
I've landed a PR upstream, but unfortunately just after the release. We might still need package.json until the next release (1.5.0 which won't be before another 6 months).

yu-re-ka replied:
mkYarnModules requires a version number in package.json, but yarnSetupHook does not

happysalada replied:
I'm glad we are getting rid of that really hard to debug error.

happysalada commented:
I really like the idea behind this PR!

One thing that always struck me when using the yarn2nix tool is the very weird layout of the result. It generates a `deps` directory and puts everything in separate directories. Going back closer to the yarn standard is really positive in my opinion.

ofborg still seems to fail, but it's probably some details. I don't have any comment about the general approach, it sounds good to me.

Happy to test when the PR is ready.

yuyuyureka committed:
plausible: use yarnSetupHook

Also changed update script to use `yarn import`.
This removes the need to create a modified package.json.

yu-re-ka commented:
I have rebased for the plausible update.

The evaluation failures are expected since I haven't converted everything away from mkYarnModules yet.

yuyuyureka committed:
rstudio: use yarnSetupHook

yuyuyureka committed:
sharedown: use yarnSetupHook

yuyuyureka committed:
emscripten: use yarnSetupHook

yuyuyureka committed:
lemmy-ui: fix build after mkYarnPackage changes

Also prevent the node_modules from being duplicated in the output.

yuyuyureka committed:
matrix-appservice-discord: fix build after mkYarnPackage changes

yuyuyureka committed:
mkYarnPackage: use yarnSetupHook

yuyuyureka committed:
uivonim: fix build after mkYarnPackage changes

yu-re-ka commented:
To-do:

- [ ] powerdns
- [ ] apache-airflow
- [ ] mastodon
- [ ] Documentation
- [ ] Release notes?

yu-re-ka commented:
Might re-do this at some point, but I don't like the current implementation anymore
